design/fast_finality.md

```diff
@@ -0,0 +1,273 @@
+limit. Rule (3) reduces the risk associated with the sequencer failing.
+
+To produce a rollup, the sequencer's host requests the creation of a rollup. The sequencer's enclave produces a rollup 
+containing all the light batches created since the last rollup, in a sparse Merkle tree structure. This rollup is 
+encrypted and signed, then sent to be included on the L1.
+
+### Discovery of rollups
+
+Nodes scan incoming L1 blocks for new rollups. They validate each new rollup by:
+
+* Checking that it is produced by the designated sequencer, based on the sequencer listed in the management contract
+* Checking that it contains all the light batches produced since the last rollup. Each light batch contains the number 
+  of the rollup that will contain it. Since the rollup is a sparse Merkle tree, proving non-inclusion of a given light 
+  batch is straightforward
+
+They then persist the rollup, so that they have a record of which light batches have been confirmed on the L1.
+
+### Resilience
+
+#### Goals
+
+The sequencer holds three important types of data:
+
+1. Transactions submitted but not yet included in a light batch
+2. Light batches (including their transactions)
+3. Rollups
+
+In the case of failover, (1) can be dropped if needed, while (3) can be recovered from the L1 chain (once submitted) or 
+recreated (provided the light batches are available). Thus, during failover, the key concern in terms of data 
+resiliency is (2).
+
+In addition, while it is acceptable to break the one-second soft-finality guarantee during failover, we should still 
+seek to minimise the recovery time. Solutions that require, for example, the full reingestion of the L1 chain are 
+unworkable.
+
+#### Cluster configuration
+
+To achieve the desired data resiliency and recovery times, the sequencer can run a cluster of `n` nodes, each backed by 
+a separate database. All the nodes are active at once. A leader node is selected to be the sole producer of light 
+batches and rollups, while the follower nodes behave like regular validator nodes, receiving the light batches via a 
+gossiping process and retrieving the rollups from the L1. Other network nodes treat each node in the cluster as a 
+regular validator node (e.g. transactions are gossiped normally, and not targeted specifically at the leader node).
+
+The cluster's leader is selected via an RPC operation on the corresponding host. It is the responsibility of the
+sequencer's operator to monitor the healthiness of the nodes. In the event that a follower crashes, it can be restarted 
+and recover data from the leader, just like a regular node. If the event that the leader crashes, the sequencer 
+operator must select a new leader.
+
+The key risk during failover to a new leader is that a single light batch (the latest) may be lost. There are two 
+specific issues that must be handled:
+
+1. Determining whether the light batch was truly lost. A new leader may come online and consider the latest light batch 
```

Comment:
I'd say that the sequencer operator can operate a vanilla cloud db instance configured with resiliency. The first thing the host does when receiving a LB from the enclave is to write it to that db instance, and wait for confirmation. Only when that is done, it begins gossiping, etc.
This ensures that LBs can't be lost.
A follower transitioning to become the leader will check that db before doing anythign.

Reply:
I guess it depends on whether we can assume the existence of an HA DB that would never lose a written light batch. Is that OK to assume? I don't have experience in that regard?

If so, I can rework this section.

Reply:
So are you saying that all the hosts connect to a single DB? Or all the hosts have separate resilient DBs, and check each others' DBs when becoming leader?

Reply:
> I guess it depends on whether we can assume the existence of an HA DB that would never lose a written light batch. Is that OK to assume? I don't have experience in that regard?
> 
> If so, I can rework this section.

yeah, we can definitely assume such a thing. Probably the safest assumption of this entire project :)